Fix exception propagation over HW exception frame on macOS arm64

janvorli opened the pull request:
There is a problem unwinding over the PAL_DispatchExceptionWrapper
to the actual hardware exception location. The unwinder is unable
to get distinct LR and PC in that frame and sets both of them to
the same value. This is caused by the fact that the
PAL_DispatchExceptionWrapper is just an injected fake frame and
there was no real call. Calls always return with LR and PC set
to the same value.

The fix unifies the hardware exception frame unwinding with Linux
where we had problems unwinding over signal handler trampoline, so
PAL_VirtualUnwind skips the trampoline and now also the
PAL_DispatchExceptionWrapper frame by copying the context of
the exception as the unwound context.

janvorli committed:
Fix exception propagation over HW exception frame on macOS arm64

There is a problem unwinding over the PAL_DispatchExceptionWrapper
to the actual hardware exception location. The unwinder is unable
to get distinct LR and PC in that frame and sets both of them to
the same value. This is caused by the fact that the
PAL_DispatchExceptionWrapper is just an injected fake frame and
there was no real call. Calls always return with LR and PC set
to the same value.

The fix unifies the hardware exception frame unwinding with Linux
where we had problems unwinding over signal handler trampoline, so
PAL_VirtualUnwind skips the trampoline and now also the
PAL_DispatchExceptionWrapper frame by copying the context of
the exception as the unwound context.

janvorli commented:
@jkotas this is an update of my previous PR that was reverted. It fixes the issue. The problem was that the context where the exception happened was updated during the unwinding - the one we were using in the `PAL_VirtualUnwind` was the one that the `PAL_SEHException` pointed to.

jkotas commented on src/tests/Regressions/coreclr/GitHub_62058/test62058.cs:
```diff
@@ -0,0 +1,60 @@
+        {
+            Console.Out.WriteLine("catch");
+            Console.Out.Flush();
+            throw new Exception("catch", e);  // throw; doesn't work either
```
```suggestion
            throw new Exception("catch", e);
```

janvorli commented:
/azp help

janvorli commented:
/azp list

janvorli commented:
There is some problem. The gcstress C tests on macos arm64 are failing massively with an assert `!CREATE_CHECK_STRING(pMT && pMT->Validate())`. I need to investigate that.

janvorli committed:
Fix GC stress C - wrong context being restored

The context that is restored in the PAL_DispatchException needs to be
the one from the exception, not the original saved one. That ensures
that the registers updated by the GC in GC stress C in the context
are properly restored after the execution is resumed.

janvorli commented:
There was a bug in the GC stress C case. In that case, the SEHProcessException returns and the context inside of the exception contains registers possibly updated by the GC. But we were resuming execution using the original captured context that doesn't have these changes. The Linux counterpart had it right, so I've just did the same thing here.

janvorli commented:
/azp run runtime-coreclr outerloop

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
@dotnet/dnceng I can see that multiple macOS x64 legs tested in this PR have failed with "No space left on device", see https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-63596-merge-65fcd545ee11487096/PayloadGroup0.Attempt.2/1/console.ebbc04fe.log?sv=2019-07-07&se=2022-01-31T23%3A39%3A01Z&sr=c&sp=rl&sig=3uWds5R3utnIIkSD7xkQtzhmHAU%2BdEk117LcqKGmzlA%3D

riarenas commented:
Looking...

riarenas commented:
@janvorli, it looks like this workitem is simply using too much disk space. The machine had 302765.5625 MB of space when the work item started, and by the time it failed, it was down to 17 MB. What is this work item doing that it uses up this much space? The machine is currently back to normal after our automatic cleanup kicked in.

janvorli commented:
@riarenas it was just a regular outerloop coreclr tests run.

riarenas commented:
I can't find any infra problems during the work item execution. My guess is there is a bug in the outerloop coreclr test runs that can end up using infinite disk space, possibly related to core dumps? Looking at our kusto logs, the `payloadgroup0` workitem seems to have uploaded almost 10 TB worth of dumps in the last 90 days so that might be something worth looking at.

janvorli committed:
Fix exception context leak in GC stress C

The PAL_SEHException had the records allocated on stack, so the
direct context restoration after the EH for GC stress C completed
leaked those.

janvorli commented:
/azp run runtime-coreclr outerloop

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

jkotas commented:
I think you can also include revert of https://github.com/dotnet/runtime/commit/39ef452edcf90bae76a1e4c8c1c24209e17ac5f2#diff-5ef2fd16f58de91bff2f05ad601ab6ff000c0febabb2312d232f20cfa132a71b in this change since it is fixing the crash.

AaronRobinsonMSFT commented on src/coreclr/pal/src/exception/machexception.cpp:
```diff
@@ -369,28 +369,22 @@ void PAL_DispatchException(PCONTEXT pContext, PEXCEPTION_RECORD pExRecord, MachE
     {
-        // The exception object takes ownership of the exceptionRecord and contextRecord
-        PAL_SEHException exception(exceptionRecord, contextRecord);
+        PAL_SEHException exception(pExRecord, pContext, true);
```
I see the `onStack` argument is being set to `true` here. This appears to be a fundamental assumption change in this code path. Is a comment warranted in this case?

janvorli replied:
The `onStack` means that we are constructing the exception from the on stack context and so when the `PAL_SEHException` is destroyed, we don't attempt to free that memory. That's why I have removed the comment above saying that it takes the ownership. I'm not sure what comment you'd like to see added here. We were constructing it from an on-heap copy of the records that were allocated right before the call before my change.

agocke commented:
The pervasive disk space issue was an artifact of GC regions causing a lot more crashes than usual.

That’s been reverted, so crashes should be more infrequent and more likely to be caused by current issues.

janvorli committed:
Reenable DllImportGenerator.Unit.Tests

am11 commented on src/coreclr/pal/src/exception/seh-unwind.cpp:
```diff
@@ -496,7 +496,9 @@ void GetContextPointers(unw_cursor_t *cursor, unw_context_t *unwContext, KNONVOL
 
```
@janvorli, should this: https://github.com/dotnet/runtime/blob/f3e4e768d10dc502b0a7e9c417ae269dcf75dbc7/src/coreclr/pal/src/exception/seh-unwind.cpp#L186-L187 be changed to:
```cpp
    fpsimd_context* fp = GetNativeSigSimdContext(unwContext);
#define ASSIGN_FP_REG(reg) if (fp) *(NEON128*) &fp->vregs[reg] = winContext->V[reg];
```
(only pass one argument to `ASSIGN_FP_REG`)
While working on #64043, I had a misconfiguration (related to `UNWIND_CONTEXT_IS_UCONTEXT_T` detection) and found this issue when it compiled on linux arm64.

AaronRobinsonMSFT replied:
@am11 Seems to be the case considering the only use of `ASSIGN_FP_REG` has a single argument.

https://github.com/dotnet/runtime/blob/f3e4e768d10dc502b0a7e9c417ae269dcf75dbc7/src/coreclr/pal/src/exception/seh-unwind.cpp#L134-L157

am11 replied:
I wonder if this code was ever compiling and is dead? cl.exe on Windows and clang on Linux both fail the `UNWIND_CONTEXT_IS_UCONTEXT_T` test and set it to 0 on `main`. To unveil the problem, i.e. misconfiguration which I had -- in order to set it to 1, delete the last `/include` from: https://github.com/dotnet/runtime/blob/f3e4e768d10dc502b0a7e9c417ae269dcf75dbc7/src/coreclr/pal/src/configure.cmake#L1033

janvorli replied:
@am11 you are right, it is a dead code. I've added it when fixing floating point registers unwind because we had that code path for ARM64. However, the libunwind currently doesn't use the ucontext_t as unwind context for ARM64, IIRC a comment in there says it is done that way in order to reduce size of the unwind context. So everything for ARM64 with `UNWIND_CONTEXT_IS_UCONTEXT_T` set is not being ever compiled.
I think that we should just remove all of that and replace it with `#error`

janvorli commented:
/backport to release/6.0
